File: src/graphtheory/BreadthFirstSearch.java
package graphtheory;

/*
 * Breadth-first search (BFS) on graph
 *
 * BFS visits nodes in order of distance away from the starting node. First, it visits nodes 1 edge away, then nodes that are
 * 2 edges away, and so on. It is not recursive, so DFS is usually better to use. BFS is used for finding the distance
 * away from the starting node for all nodes in an unweighted graph.
 *
 * Time complexity: O(V + E) , where V is the number of vertices and E is the number of edges
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class BreadthFirstSearch {
    static ArrayList<Integer>[] adj; // Represent graph using Adjacency List
    static int[] dist; // Represents distance from each node to starting node

    static void bfs(int k) {
        Arrays.fill(dist, -1); // fill distance array with -1's to denote that these nodes have not been processed yet
        Queue<Integer> q = new LinkedList<>();
        dist[k] = 0;
        q.add(k);
        while(!q.isEmpty()) {
            int v = q.poll();
            for(int e : adj[v]) {
                if(dist[e] == -1) {
                    dist[e] = dist[v] + 1;
                    q.add(e);
                }
            }
        }
    }
}

File: src/c++/graph/BreadthFirstSearch.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

void bfs(vector<vector<int>>& graph, vector<bool> visited, int startNode);

/*
 * BREADTH FIRST SEARCH (DFS) ALGORITHM
 *
 * A graph traversal algorithm that explores all nodes at the present depth before moving on to nodes at the next depth level by using a queue
 *
 * Time complexity: O(V + E)
 *
 */

void bfs(vector<vector<int>>& graph, vector<bool> visited, int startNode) {
    queue<int> queue;

    queue.push(startNode);
    visited[startNode] = true;
    while (!queue.empty()) {
        int c = queue.front();
        queue.pop();
        cout << c << " ";
        for (int neighbor : graph[c]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                queue.push(neighbor);
            }
        }
    }
}

int main() {
    vector<vector<int>> graph = {
            {1, 2}, // Node 0 is connected to nodes 1 and 2
            {0, 3}, // Node 1 is connected to nodes 0 and 3
            {0, 4}, // Node 2 is connected to nodes 0 and 4
            {1},    // Node 3 is connected to node 1
            {2}     // Node 4 is connected to node 2
    };
    vector<bool> visited(graph.size());

    bfs(graph, visited,0); // 0 1 2 3 4
    cout << endl;

    return 0;
}